src/Policies.cpp

```diff
@@ -96,17 +96,79 @@ std::vector<Policy *> PolicyManager::load(const YAML::Node &policy_yaml)
+        if (input_node["tap"] && input_node["tap_selector"]) {
+            throw PolicyException("input can have only key 'input.tap' or key 'input.tap_selector'");
+        } else if (!input_node["tap"] && !input_node["tap_selector"]) {
+            throw PolicyException("missing key 'input.tap' or key 'input.tap_selector'");
+        }
+
+        if (input_node["tap"]) {
+            if (!input_node["tap"].IsScalar()) {
+                throw PolicyException("invalid tap at key 'input.tap'");
+            }
+            result.push_back(create_policy(it->second, policy_name));
+        } else if (input_node["tap_selector"]) {
+            auto tap_selector = input_node["tap_selector"];
+            if (!tap_selector.IsMap()) {
+                throw PolicyException("'input.tap_selector' is not a map");
+            }
+
+            if (tap_selector["all"] && tap_selector["any"]) {
+                throw PolicyException("input can have only key 'input.tap_selector.all' or key 'input.tap_selector.any'");
+            } else if (!tap_selector["all"] && !tap_selector["any"]) {
+                throw PolicyException("missing key 'input.tap_selector.all' or key 'input.tap_selector.any'");
+            }
+
+            bool all = true;
+            std::string binary_op;
+            if (tap_selector["all"]) {
+                if (!tap_selector["all"].IsMap()) {
+                    throw PolicyException("'input.tap_selector.all' is not a map");
+                }
+                all = true;
+                binary_op = "all";
+            } else if (tap_selector["any"] || tap_selector["any"].IsMap()) {
+                if (!tap_selector["any"].IsMap()) {
+                    throw PolicyException("'input.tap_selector.any' is not a map");
+                }
+                all = false;
+                binary_op = "any";
+            }
+
+            bool match = false;
+            auto [tap_modules, hm_lock] = _registry->tap_manager()->module_get_all_locked();
+            for (auto &[name, mod] : tap_modules) {
+                auto tmod = dynamic_cast<Tap *>(mod.get());
+                if (tmod && tmod->tags_validate_yaml(tap_selector[binary_op], all)) {
+                    auto policy_selector_name = policy_name + "_" + tmod->name();
+                    if (module_exists(policy_selector_name)) {
+                        throw PolicyException(fmt::format("policy with name '{}' already defined", policy_selector_name));
+                    }
+                    result.push_back(create_policy(it->second, policy_selector_name, tmod));
+                    match = true;
+                }
+            }
+            if (!match) {
+                throw PolicyException("no tap match found for specified 'input.tap_selector' tags");
+            }
+        }
+    }
+
+    return result;
+}
+
+Policy *PolicyManager::create_policy(const YAML::Node &policy_yaml, std::string policy_name, Tap *tap)
```

Comment:
I like that you broke this out, but wonder if we can break it out even more - it's still very large, hard to understand method

Reply:
I break out the handler validation as well